content/post/73-putting-npm-on-ipfs-part-1.md

```diff
@@ -0,0 +1,122 @@
+      "dist": {
+        "tarball": "https://registry.js.ipfs.io/ipfs/-/ipfs-0.34.4.tgz",
+        "source": "https://registry.npmjs.org/ipfs/-/ipfs-0.34.4.tgz",
+        "cid": "bafybeiafts7s65iodk4wsetucwhury3cpk4fge374wxhdu5vzc4zuli4xi"
+      }
+    },
+...
+```
+
+The CID resolves to the tarball for a module, but while the module has been added to IPFS is not stored in the MFS so we resolve it using the CID:
+
+```console
+$ jsipfs files read /ipfs/bafybeiafts7s65iodk4wsetucwhury3cpk4fge374wxhdu5vzc4zuli4xi
+??V?0:O=E??g?ܢ?]n<?l?...
+```
+
+### 🔍 Resolving a tarball
+
+The npm CLI first requests the metadata for a given module - this contains all published versions and URLs to tarballs for that version.  It is a JSON document the CLI uses to select which version to request that satisfies the [semver](https://semver.org/) range the developer has specified in their `package.json` file.
+
+The CLI then requests a tarball - we use the URL to load the metadata from the package and from that to look up the CID of the tarball.  If the metadata contains a CID for the package, we use it to fetch the content from the IPFS network and relay it to the user.  If the CID is not present, we request the tarball from the npm registry and add it to IPFS while streaming it back to the user. Once the stream is complete we update the manifest with the CID for future use.
+
+![Request sequence](/73-putting-npm-on-ipfs-part-1/request.png)
+
+### 🆕 Updates
+
+What about new modules? npm is constantly being updated with people publishing new modules every few minutes. npm [publishes a feed](https://replicate.npmjs.com/registry) that works like a CouchDB replication log - we have a [replication server](https://github.com/ipfs-shipyard/ipfs-npm-registry-mirror/tree/master/packages/replication-master) that watches this feed for new modules being published.
+
+![Replication](/73-putting-npm-on-ipfs-part-1/replication.png)
+
+When it sees a new module published, the replicator pulls down any new tarballs, adds them to IPFS and updates the metadata with the CIDs of the new module.
+
+It then uses [pubsub](https://blog.ipfs.io/25-pubsub/) to inform the mirrors of the update, which then update their local copies of the module metadata with the new versions & CIDs.
+
+### 🧳 Host your own version
+
+The code for [ipfs-npm-registry-mirror](https://github.com/ipfs-shipyard/ipfs-npm-registry-mirror) is completely open source so you can host a version on your own infrastructure, or on a Raspberry Pi. If you try this, [let us know](https://github.com/ipfs-shipyard/ipfs-npm-registry-mirror/issues) how you get on!
+
+## 🎁 What's next?
+
+While you can use the registry today by specifying the `--registry=https://registry.js.ipfs.io` flag, you are still using HTTP to request modules even if they are then resolved internally via IPFS.
+
+What if you could skip that part and just use IPFS?
+
+Stay tuned for [Putting npm on IPFS Part 2 - The Client](/post/74-putting-npm-on-ipfs-part-2)
```

Comment:
Can't wait. Can we get this one merged and published 🚢🚢🚢??

Reply:
@achingbrain can has final review and finishing touches pls?